File: pixer_client.py
import socket
import time
import logging
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)

class PixerConnectionError(Exception):
    """Custom exception for Pixer device connection errors"""
    pass

class PixerClient:
    """
    Client for communicating with Pixer E-Ink photo frame device.
    Handles TCP connection to device at 192.168.1.1:6000
    """
    
    def __init__(self, host: str = '192.168.1.1', port: int = 6000):
        self.host = host
        self.port = port
        self.socket = None
        self.input = None
        self.output = None
        self.is_connected = False

    def connect(self, max_attempts: int = 10, retry_delay: float = 2.0) -> bool:
        """
        Establish connection to Pixer device
        
        Args:
            max_attempts: Maximum number of connection attempts
            retry_delay: Delay between connection attempts in seconds
            
        Returns:
            bool: True if connection successful, False otherwise
            
        Raises:
            PixerConnectionError: If connection fails after all attempts
        """
        for attempt in range(max_attempts):
            try:
                self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                self.socket.settimeout(2)
                self.socket.connect((self.host, self.port))
                self.input = self.socket.makefile('rb')
                self.output = self.socket.makefile('wb')
                self.is_connected = True
                logger.debug(f"Connected to Pixer device at {self.host}:{self.port}")
                return True
            except Exception as e:
                logger.debug(f"Connection attempt {attempt + 1}/{max_attempts} failed: {e}")
                if self.socket:
                    self.socket.close()
                    self.socket = None
                if attempt < max_attempts - 1:
                    time.sleep(retry_delay)
        
        self.is_connected = False
        raise PixerConnectionError(f"Failed to connect to Pixer device after {max_attempts} attempts")

    def send_command(self, command: bytes, timeout_retries: int = 5) -> Optional[str]:
        """
        Send a command to the Pixer device and receive response
        
        Args:
            command: Command to send as bytes
            timeout_retries: Number of timeout retries
            
        Returns:
            str: Response from device, or None if failed
        """
        if not self.is_connected or not self.socket:
            raise PixerConnectionError("Not connected to device")
            
        try:
            self.output.write(command)
            self.output.flush()
            
            for retry in range(timeout_retries):
                try:
                    chunk = self.socket.recv(64)
                    if chunk:
                        response = chunk.decode().strip()
                        logger.debug(f"Command: {command}, Response: {response}")
                        return response
                except socket.timeout:
                    logger.debug(f"Timeout on retry {retry + 1}/{timeout_retries}")
                    
            logger.warning(f"No response received for command: {command}")
            return None
            
        except Exception as e:
            logger.error(f"Error sending command {command}: {e}")
            return None

    def test_connection(self) -> bool:
        """
        Test if device is responding correctly
        
        Returns:
            bool: True if device responds with expected message
        """
        try:
            response = self.send_command(b"#TEST#")
            return response == "Hello PC!"
        except Exception as e:
            logger.error(f"Connection test failed: {e}")
            return False

    def get_device_info(self) -> Dict[str, Any]:
        """
        Get comprehensive device information
        
        Returns:
            dict: Device information including versions and battery level
        """
        info = {
            'connected': False,
            'battery_level': None,
            'ble_version': None,
            'ite_version': None,
            'mcu_version': None,
            'error': None
        }
        
        try:
            if not self.test_connection():
                info['error'] = "Device not responding to test command"
                return info
                
            info['connected'] = True
            info['battery_level'] = self.send_command(b"batteryLevel")
            info['ble_version'] = self.send_command(b"bleVersion")
            info['ite_version'] = self.send_command(b"iteVersion")
            info['mcu_version'] = self.send_command(b"mcuVersion")
            
            # Convert battery level to integer if possible
            if info['battery_level']:
                try:
                    info['battery_level'] = int(info['battery_level'])
                except ValueError:
                    pass
                    
        except Exception as e:
            info['error'] = str(e)
            logger.error(f"Error getting device info: {e}")
            
        return info

    def reset_device(self) -> bool:
        """
        Send reset command to device
        
        Returns:
            bool: True if reset command sent successfully
        """
        try:
            if not self.test_connection():
                return False
                
            response = self.send_command(b"reset")
            logger.debug("Reset command sent to device")
            return True
            
        except Exception as e:
            logger.error(f"Error resetting device: {e}")
            return False

    def upload_image_data(self, data: bytes, chunk_size: int = 4096) -> Dict[str, Any]:
        """
        Upload image data to the Pixer device
        
        Args:
            data: Image data to upload
            chunk_size: Size of chunks to send
            
        Returns:
            dict: Upload result with success status and progress info
        """
        result = {
            'success': False,
            'bytes_sent': 0,
            'total_bytes': len(data),
            'error': None
        }
        
        try:
            if not self.socket:
                raise PixerConnectionError("Not connected to device")
                
            self.socket.settimeout(10)
            offset = 0
            
            while offset < len(data):
                chunk = data[offset:offset + chunk_size]
                self.socket.sendall(chunk)
                offset += len(chunk)
                result['bytes_sent'] = offset
                
                # Log progress every 10%
                progress = (offset * 100) // len(data)
                if progress % 10 == 0:
                    logger.debug(f"Upload progress: {progress}%")
            
            # Send completion marker
            tail = "#MOVE#d"
            self.socket.sendall(tail.encode('utf-8'))
            
            result['success'] = True
            logger.debug(f"Successfully uploaded {len(data)} bytes")
            
        except Exception as e:
            result['error'] = str(e)
            logger.error(f"Error uploading image data: {e}")
            
        return result

    def close(self):
        """Close connection to device"""
        self.is_connected = False
        if self.socket:
            try:
                self.socket.close()
            except Exception as e:
                logger.debug(f"Error closing socket: {e}")
            finally:
                self.socket = None
                self.input = None
                self.output = None
        logger.debug("Connection closed")

    def __enter__(self):
        """Context manager entry"""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.close()

File: src/pixer-client.js
const net = require('net');
const { EventEmitter } = require('events');

class PixerClient extends EventEmitter {
    constructor(host = '192.168.1.1', port = 6000) {
        super();
        this.host = host;
        this.port = port;
        this.socket = null;
        this.isConnected = false;
    }

    async connect() {
        return new Promise((resolve, reject) => {
            let attempts = 0;
            const maxAttempts = 10;
            
            const tryConnect = () => {
                attempts++;
                console.log(`Connection attempt ${attempts}/${maxAttempts}`);
                
                this.socket = new net.Socket();
                this.socket.setTimeout(2000);
                
                this.socket.on('connect', () => {
                    console.log('Connected successfully');
                    this.isConnected = true;
                    resolve();
                });
                
                this.socket.on('error', (error) => {
                    console.log(`Connection attempt ${attempts} failed: ${error.message}`);
                    this.socket.destroy();
                    
                    if (attempts < maxAttempts) {
                        setTimeout(tryConnect, 2000);
                    } else {
                        this.isConnected = false;
                        reject(new Error(`Failed to connect after ${maxAttempts} attempts`));
                    }
                });
                
                this.socket.on('timeout', () => {
                    console.log(`Connection attempt ${attempts} timed out`);
                    this.socket.destroy();
                    
                    if (attempts < maxAttempts) {
                        setTimeout(tryConnect, 2000);
                    } else {
                        this.isConnected = false;
                        reject(new Error(`Connection timed out after ${maxAttempts} attempts`));
                    }
                });
                
                this.socket.connect(this.port, this.host);
            };
            
            tryConnect();
        });
    }

    async send(data) {
        return new Promise((resolve, reject) => {
            if (!this.socket || !this.isConnected) {
                reject(new Error('Not connected to device'));
                return;
            }

            let responseReceived = false;
            let retries = 0;
            const maxRetries = 5;

            const handleResponse = (chunk) => {
                if (!responseReceived) {
                    responseReceived = true;
                    this.socket.removeListener('data', handleResponse);
                    resolve(chunk.toString());
                }
            };

            const handleTimeout = () => {
                retries++;
                console.log(`Timeout, retrying... ${retries}/${maxRetries}`);
                
                if (retries < maxRetries) {
                    this.socket.setTimeout(2000);
                } else {
                    this.socket.removeListener('data', handleResponse);
                    if (!responseReceived) {
                        reject(new Error('Response timeout after multiple retries'));
                    }
                }
            };

            this.socket.on('data', handleResponse);
            this.socket.on('timeout', handleTimeout);
            this.socket.setTimeout(2000);

            try {
                this.socket.write(data);
            } catch (error) {
                this.socket.removeListener('data', handleResponse);
                reject(new Error(`Error sending data: ${error.message}`));
            }
        });
    }

    async uploadImage(data, progressCallback) {
        return new Promise(async (resolve, reject) => {
            try {
                // Ensure we have a fresh connection for upload
                await this.connect();

                // Verify connection with test command
                const testResponse = await this.send(Buffer.from('#TEST#'));
                if (testResponse !== 'Hello PC!') {
                    throw new Error('Invalid test response from device');
                }

                this.socket.setTimeout(10000);
                const chunkSize = 4096;
                let offset = 0;

                const sendChunk = () => {
                    if (offset >= data.length) {
                        // Send the tail command
                        const tail = Buffer.from('#MOVE#d', 'utf-8');
                        this.socket.write(tail);
                        this.close();
                        resolve('Upload completed');
                        return;
                    }

                    const chunk = data.slice(offset, offset + chunkSize);
                    this.socket.write(chunk);
                    offset += chunk.length;

                    const progress = Math.floor((offset * 100) / data.length);
                    if (progressCallback) {
                        progressCallback(progress);
                    }

                    // Continue with next chunk
                    setImmediate(sendChunk);
                };

                sendChunk();

            } catch (error) {
                this.close();
                reject(new Error(`Error in upload: ${error.message}`));
            }
        });
    }

    async checkDevice() {
        try {
            await this.connect();
            
            const testResponse = await this.send(Buffer.from('#TEST#'));
            if (testResponse !== 'Hello PC!') {
                throw new Error('Invalid test response from device');
            }
            
            const bleVersion = await this.send(Buffer.from('bleVersion'));
            const iteVersion = await this.send(Buffer.from('iteVersion'));
            const mcuVersion = await this.send(Buffer.from('mcuVersion'));
            const batteryLevel = await this.send(Buffer.from('batteryLevel'));
            
            const deviceInfo = {
                bleVersion: bleVersion.trim(),
                iteVersion: iteVersion.trim(),
                mcuVersion: mcuVersion.trim(),
                batteryLevel: parseInt(batteryLevel.trim()) || 0
            };
            
            this.close();
            return deviceInfo;
            
        } catch (error) {
            this.close();
            throw error;
        }
    }

    async resetDevice() {
        try {
            await this.connect();
            
            const testResponse = await this.send(Buffer.from('#TEST#'));
            if (testResponse !== 'Hello PC!') {
                throw new Error('Invalid test response from device');
            }
            
            await this.send(Buffer.from('reset'));
            this.close();
            
        } catch (error) {
            this.close();
            throw error;
        }
    }

    close() {
        if (this.socket) {
            this.socket.destroy();
            this.socket = null;
        }
        this.isConnected = false;
    }
}

module.exports = PixerClient;
